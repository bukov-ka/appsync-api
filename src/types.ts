export interface Order {
  id: string;
  date: string;
  totalAmount: number;
  email: string;
  customer: Customer;
  product: string;
  products: Product[];
}


export interface Customer {
  email: string;
  fullName: string;
}

export interface Product {
  name: string;
  price: number;
  quantity: number;
}

export interface LambdaEvent {
  info: {
    fieldName: string;
  };
  arguments: {
    email: string;
    orderDate: string;
  };
}
