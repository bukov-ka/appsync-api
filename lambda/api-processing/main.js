const AWS = require('aws-sdk');
const { v4: uuidv4 } = require('uuid');

const dynamoDb = new AWS.DynamoDB.DocumentClient();
const CUSTOMER_TABLE = process.env.CUSTOMER_TABLE;
const ORDER_TABLE = process.env.ORDER_TABLE;
const PRODUCT_TABLE = process.env.PRODUCT_TABLE;

exports.handler = async (event) => {
  switch (event.info.fieldName) {
    case 'orders':
      console.log(`event: ${JSON.stringify(event)}`);
      return await getOrders(event.arguments.email, event.arguments.orderDate);
    default:
      throw new Error(`Unknown field "${event.info.fieldName}"`);
  }
};

async function getOrders(email, orderDate) {
  const params = {
    TableName: ORDER_TABLE,
    KeyConditionExpression: 'email = :e AND #orderDate = :d',
    ExpressionAttributeValues: {
      ':e': email,
      ':d': orderDate,
    },
    ExpressionAttributeNames: {
      '#orderDate': 'date', // Alias for the reserved keyword 'date'
    },
  };


  try {
    const result = await dynamoDb.query(params).promise();

    // Fetch additional data for nested fields (such as customer and products)
    const orders = await Promise.all(result.Items.map(async (order) => ({
      ...order,
      customer: await getCustomerByEmail(order.email),
      products: await getProductsByOrderId(order.id),
    })));

    return orders;
  } catch (error) {
    console.error('getOrders:' + error, params);
    throw error;
  }
}


// Function to fetch a customer by email
async function getCustomerByEmail(email) {
  const params = {
    TableName: CUSTOMER_TABLE,
    Key: {
      email,
    },
  };

  try {
    const result = await dynamoDb.get(params).promise();
    return result.Item || null;
  } catch (error) {
    console.error('getCustomerByEmail:' + error, params);
    throw error;
  }
}

// Function to fetch products for each order
async function getProductsByOrderId(orderId) {
  // Placeholder: You need to adjust the following code according to your data model to link products with orders.
  const params = {
    TableName: PRODUCT_TABLE,
  };

  try {
    const result = await dynamoDb.scan(params).promise();
    return result.Items || [];
  } catch (error) {
    console.error('getProductsByOrderId:' + error, params);
    throw error;
  }
}
