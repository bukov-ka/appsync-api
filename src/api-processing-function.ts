// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import * as path from 'path';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

/**
 * Props for ApiProcessingFunction
 */
export interface ApiProcessingFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/api-processing.
 */
export class ApiProcessingFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: ApiProcessingFunctionProps) {
    super(scope, id, {
      description: 'src/api-processing.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs16.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/api-processing.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}