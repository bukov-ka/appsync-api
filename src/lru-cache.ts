// Define LRUCache class with private properties capacity, ttl, and cache
export class LRUCache {
  // Capacity of the cache
  private capacity: number;
  // Time-to-live (TTL) in milliseconds
  private ttl: number;
  // Cache map that stores key-value pairs with a timestamp
  private cache: Map<string, { timestamp: number; value: any }>;

  // Constructor takes capacity and ttl as arguments and initializes the cache map
  constructor(capacity: number, ttl: number) {
    this.capacity = capacity;
    this.ttl = ttl;
    this.cache = new Map();
  }

  // Get method takes a key and returns the cached value or null if not found or expired
  get(key: string): any | null {
    const cachedItem = this.cache.get(key);
    // If cached item is not found or it's expired, return null
    if (!cachedItem || Date.now() - cachedItem.timestamp > this.ttl) {
      return null;
    }
    // Update the timestamp of the cached item to mark it as recently used
    cachedItem.timestamp = Date.now();
    return cachedItem.value;
  }

  // Set method takes a key and value and stores it in the cache
  set(key: string, value: any): void {
    // If cache is at capacity, remove the least recently used (LRU) item
    if (this.cache.size >= this.capacity) {
      const lruKey = Array.from(this.cache.keys())[0];
      this.cache.delete(lruKey);
    }
    // Add the new item to the cache with a timestamp
    this.cache.set(key, { timestamp: Date.now(), value });
  }
}
