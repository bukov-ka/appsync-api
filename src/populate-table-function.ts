// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import * as path from 'path';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

/**
 * Props for PopulateTableFunction
 */
export interface PopulateTableFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/populate-table.
 */
export class PopulateTableFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: PopulateTableFunctionProps) {
    super(scope, id, {
      description: 'src/populate-table.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs16.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/populate-table.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}