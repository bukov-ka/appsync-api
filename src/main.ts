import { App, Stack, StackProps } from 'aws-cdk-lib';
import * as awsCdk from 'aws-cdk-lib';
import * as appsync from 'aws-cdk-lib/aws-appsync';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as customResources from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';
// Import the generated lambdas
import { ApiProcessingFunction } from './api-processing-function';
import { PopulateTableFunction } from './populate-table-function';

const API_KEY_EXPIRATION = awsCdk.Duration.days(365);

export class OrdersAPIStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    // Create the DynamoDB tables
    const { orderTable, productTable } = this.createDynamoDbTableStructure();

    // Create a custom resource to trigger the Lambda function on stack deployment
    this.populateDataOnDeploy(orderTable, productTable);

    // Create the AppSync API
    const appSyncAPI = new appsync.GraphqlApi(this, 'Api', {
      name: 'customerOrders',
      schema: appsync.SchemaFile.fromAsset('graphql/schema.graphql'),
      authorizationConfig: {
        defaultAuthorization: {
          authorizationType: appsync.AuthorizationType.API_KEY,
          apiKeyConfig: {
            expires: awsCdk.Expiration.after(API_KEY_EXPIRATION),
          },
        },
      },
      xrayEnabled: true,
    });

    // Lambda to process requests
    const backendLambda = new ApiProcessingFunction(this, 'AppSyncHandler');
    // Grant read permissions on DynamoDB table to the Lambda function
    orderTable.grantReadData(backendLambda);
    productTable.grantReadData(backendLambda);
    backendLambda.addEnvironment('ORDER_TABLE', orderTable.tableName);
    backendLambda.addEnvironment('PRODUCT_TABLE', productTable.tableName);

    // Set the data source for the API
    const customerDataSource = appSyncAPI.addLambdaDataSource('CustomerTableDataSource', backendLambda);

    // Create resolvers
    customerDataSource.createResolver(
      'customerQueryResolver',
      {
        typeName: 'Query',
        fieldName: 'orders',
        requestMappingTemplate: appsync.MappingTemplate.lambdaRequest(),
        responseMappingTemplate: appsync.MappingTemplate.lambdaResult(),
      },
    );

    // Output the API Key and URL
    new awsCdk.CfnOutput(this, 'GraphQLAPIURL', {
      value: appSyncAPI.graphqlUrl,
    });

    new awsCdk.CfnOutput(this, 'GraphQLAPIKey', {
      value: appSyncAPI.apiKey || '',
    });
  }

  // Create the DynamoDB table for Customers, Orders, Products
  private createDynamoDbTableStructure() {
    // Create the DynamoDB table for Orders
    const orderTable = new dynamodb.Table(this, 'orderTable', {
      partitionKey: { name: 'lineId', type: dynamodb.AttributeType.STRING },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
    });

    // Add GSI for querying by email and date
    orderTable.addGlobalSecondaryIndex({
      indexName: 'emailDateIndex',
      partitionKey: { name: 'email', type: dynamodb.AttributeType.STRING },
      sortKey: { name: 'date', type: dynamodb.AttributeType.STRING },
    });

    // Create the DynamoDB table for Products
    const productTable = new dynamodb.Table(this, 'productTable', {
      partitionKey: { name: 'id', type: dynamodb.AttributeType.STRING }, // Change partitionKey from 'name' to 'id'
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
    });

    // This is a demo app, so we'll clean up resources when the stack is deleted
    // For production we need to retain the data
    orderTable.applyRemovalPolicy(awsCdk.RemovalPolicy.DESTROY);
    productTable.applyRemovalPolicy(awsCdk.RemovalPolicy.DESTROY);
    return { orderTable, productTable };
  }

  // Create a custom resource to trigger the Lambda function on stack deployment
  private populateDataOnDeploy(orderTable: dynamodb.Table, productTable: dynamodb.Table) {
    // Create a new Lambda function to populate the table
    const populateTableFunction = new PopulateTableFunction(this, 'PopulateTableFunction');

    // Grant read and write access to your DynamoDB table for the Lambda function
    orderTable.grantReadWriteData(populateTableFunction);
    productTable.grantReadWriteData(populateTableFunction);

    // Add the table name as an environment variable to the Lambda function
    populateTableFunction.addEnvironment('ORDER_TABLE', orderTable.tableName);
    populateTableFunction.addEnvironment('PRODUCT_TABLE', productTable.tableName);
    // Create a custom role to add policies to it
    const customResourceRole = new iam.Role(this, 'CustomResourceRole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
    });

    customResourceRole.addToPolicy(new iam.PolicyStatement({
      actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
      resources: ['arn:aws:logs:*:*:*'],
    }));

    customResourceRole.addToPolicy(new iam.PolicyStatement({
      actions: ['lambda:InvokeFunction'],
      resources: [populateTableFunction.functionArn],
    }));

    new customResources.AwsCustomResource(this, 'PopulateTableCustomResource', {
      onCreate: {
        service: 'Lambda',
        action: 'invoke',
        physicalResourceId: customResources.PhysicalResourceId.of('PopulateTableCustomResourceId'),
        parameters: {
          FunctionName: populateTableFunction.functionName,
        },
      },
      policy: customResources.AwsCustomResourcePolicy.fromSdkCalls({ resources: customResources.AwsCustomResourcePolicy.ANY_RESOURCE }),
      role: customResourceRole,
    });
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new OrdersAPIStack(app, 'appsync-api-dev', { env: devEnv });
// new MyStack(app, 'appsync-api-prod', { env: prodEnv });

app.synth();