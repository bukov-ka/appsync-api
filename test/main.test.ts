import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { OrdersAPIStack } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new OrdersAPIStack(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();

  // Check for a Lambda function 'Populate*'
  const populateLambda = Object.keys(template.findResources('AWS::Lambda::Function')).filter(
    (key) => key.startsWith('Populate'),
  );
  expect(populateLambda.length).toBeGreaterThan(0);

  // Check that there is at least one DynamoDB table
  template.resourceCountIs('AWS::DynamoDB::Table', 2);

  // Check for an AppSync API
  template.resourceCountIs('AWS::AppSync::GraphQLApi', 1);

  // Check for GraphQLAPIKey and GraphQLAPIURL strings in the output
  expect(Object.keys(template.findOutputs('GraphQLAPIKey')).length).toBeGreaterThan(0);
  expect(Object.keys(template.findOutputs('GraphQLAPIURL')).length).toBeGreaterThan(0);
});
