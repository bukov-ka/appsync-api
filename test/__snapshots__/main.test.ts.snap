// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Mappings": {
    "DefaultCrNodeVersionMap": {
      "af-south-1": {
        "value": "nodejs16.x",
      },
      "ap-east-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-2": {
        "value": "nodejs16.x",
      },
      "ap-northeast-3": {
        "value": "nodejs16.x",
      },
      "ap-south-1": {
        "value": "nodejs16.x",
      },
      "ap-south-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-1": {
        "value": "nodejs16.x",
      },
      "ap-southeast-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-3": {
        "value": "nodejs16.x",
      },
      "ca-central-1": {
        "value": "nodejs16.x",
      },
      "cn-north-1": {
        "value": "nodejs16.x",
      },
      "cn-northwest-1": {
        "value": "nodejs16.x",
      },
      "eu-central-1": {
        "value": "nodejs16.x",
      },
      "eu-central-2": {
        "value": "nodejs16.x",
      },
      "eu-north-1": {
        "value": "nodejs16.x",
      },
      "eu-south-1": {
        "value": "nodejs16.x",
      },
      "eu-south-2": {
        "value": "nodejs16.x",
      },
      "eu-west-1": {
        "value": "nodejs16.x",
      },
      "eu-west-2": {
        "value": "nodejs16.x",
      },
      "eu-west-3": {
        "value": "nodejs16.x",
      },
      "me-central-1": {
        "value": "nodejs16.x",
      },
      "me-south-1": {
        "value": "nodejs16.x",
      },
      "sa-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-2": {
        "value": "nodejs16.x",
      },
      "us-gov-east-1": {
        "value": "nodejs16.x",
      },
      "us-gov-west-1": {
        "value": "nodejs16.x",
      },
      "us-iso-east-1": {
        "value": "nodejs14.x",
      },
      "us-iso-west-1": {
        "value": "nodejs14.x",
      },
      "us-isob-east-1": {
        "value": "nodejs14.x",
      },
      "us-west-1": {
        "value": "nodejs16.x",
      },
      "us-west-2": {
        "value": "nodejs16.x",
      },
    },
  },
  "Outputs": {
    "GraphQLAPIKey": {
      "Value": {
        "Fn::GetAtt": [
          "ApiDefaultApiKeyF991C37B",
          "ApiKey",
        ],
      },
    },
    "GraphQLAPIURL": {
      "Value": {
        "Fn::GetAtt": [
          "ApiF70053CD",
          "GraphQLUrl",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "CustomResourceRoleDefaultPolicyFA36F832",
        "CustomResourceRoleAB1EF463",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "bb459fac5f6b4b052aac9803443226d161a5cfe96f4648b21f9e4912c698bf30.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "CustomResourceRoleAB1EF463",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "DefaultCrNodeVersionMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ApiCustomerTableDataSource8154D467": {
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "ApiF70053CD",
            "ApiId",
          ],
        },
        "LambdaConfig": {
          "LambdaFunctionArn": {
            "Fn::GetAtt": [
              "AppSyncHandlerC41B9ACB",
              "Arn",
            ],
          },
        },
        "Name": "CustomerTableDataSource",
        "ServiceRoleArn": {
          "Fn::GetAtt": [
            "ApiCustomerTableDataSourceServiceRole3DFE90E6",
            "Arn",
          ],
        },
        "Type": "AWS_LAMBDA",
      },
      "Type": "AWS::AppSync::DataSource",
    },
    "ApiCustomerTableDataSourceServiceRole3DFE90E6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "appsync.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ApiCustomerTableDataSourceServiceRoleDefaultPolicy83A56B52": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "AppSyncHandlerC41B9ACB",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "AppSyncHandlerC41B9ACB",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ApiCustomerTableDataSourceServiceRoleDefaultPolicy83A56B52",
        "Roles": [
          {
            "Ref": "ApiCustomerTableDataSourceServiceRole3DFE90E6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ApiDefaultApiKeyF991C37B": {
      "DependsOn": [
        "ApiSchema510EECD7",
      ],
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "ApiF70053CD",
            "ApiId",
          ],
        },
        "Expires": 1716609197,
      },
      "Type": "AWS::AppSync::ApiKey",
    },
    "ApiF70053CD": {
      "Properties": {
        "AuthenticationType": "API_KEY",
        "Name": "customerOrders",
        "XrayEnabled": true,
      },
      "Type": "AWS::AppSync::GraphQLApi",
    },
    "ApiSchema510EECD7": {
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "ApiF70053CD",
            "ApiId",
          ],
        },
        "Definition": "type Customer {
  email: String!
  fullName: String!
}

type Order {
  id: ID!
  date: String!
  totalAmount: Float!
  customer: Customer!
  products: [Product!]!
}

type Product {
  name: String!
  price: Float!
  quantity: Int!
}

type Query {
  orders(email: String!, orderDate: String!): [Order!]!
}
",
      },
      "Type": "AWS::AppSync::GraphQLSchema",
    },
    "ApicustomerQueryResolver00117096": {
      "DependsOn": [
        "ApiCustomerTableDataSource8154D467",
        "ApiSchema510EECD7",
      ],
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "ApiF70053CD",
            "ApiId",
          ],
        },
        "DataSourceName": "CustomerTableDataSource",
        "FieldName": "orders",
        "Kind": "UNIT",
        "RequestMappingTemplate": "{"version": "2017-02-28", "operation": "Invoke", "payload": $util.toJson($ctx)}",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Query",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "AppSyncHandlerC41B9ACB": {
      "DependsOn": [
        "AppSyncHandlerServiceRoleDefaultPolicy78985631",
        "AppSyncHandlerServiceRole9FC2FD31",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e4d27837937937e3e7102f65da3e750a35b9ef1d3de463986fc6ef92a007c535.zip",
        },
        "Description": "src/api-processing.lambda.ts",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "ORDER_TABLE": {
              "Ref": "orderTable2AB11F49",
            },
            "PRODUCT_TABLE": {
              "Ref": "productTable02AC15CB",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AppSyncHandlerServiceRole9FC2FD31",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "AppSyncHandlerServiceRole9FC2FD31": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AppSyncHandlerServiceRoleDefaultPolicy78985631": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "orderTable2AB11F49",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "orderTable2AB11F49",
                          "Arn",
                        ],
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "productTable02AC15CB",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AppSyncHandlerServiceRoleDefaultPolicy78985631",
        "Roles": [
          {
            "Ref": "AppSyncHandlerServiceRole9FC2FD31",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomResourceRoleAB1EF463": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomResourceRoleDefaultPolicyFA36F832": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:logs:*:*:*",
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "PopulateTableFunctionF6347F15",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomResourceRoleDefaultPolicyFA36F832",
        "Roles": [
          {
            "Ref": "CustomResourceRoleAB1EF463",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PopulateTableCustomResourceCustomResourcePolicy9B9419F6": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:Invoke",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PopulateTableCustomResourceCustomResourcePolicy9B9419F6",
        "Roles": [
          {
            "Ref": "CustomResourceRoleAB1EF463",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PopulateTableCustomResourceF403D384": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "PopulateTableCustomResourceCustomResourcePolicy9B9419F6",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"Lambda","action":"invoke","physicalResourceId":{"id":"PopulateTableCustomResourceId"},"parameters":{"FunctionName":"",
              {
                "Ref": "PopulateTableFunctionF6347F15",
              },
              ""}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "PopulateTableFunctionF6347F15": {
      "DependsOn": [
        "PopulateTableFunctionServiceRoleDefaultPolicy7C543758",
        "PopulateTableFunctionServiceRoleB44223E7",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "396ea4edf2db369eddf8baa9dac6b43b4f643e01461271624caf78468613144a.zip",
        },
        "Description": "src/populate-table.lambda.ts",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "ORDER_TABLE": {
              "Ref": "orderTable2AB11F49",
            },
            "PRODUCT_TABLE": {
              "Ref": "productTable02AC15CB",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "PopulateTableFunctionServiceRoleB44223E7",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "PopulateTableFunctionServiceRoleB44223E7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PopulateTableFunctionServiceRoleDefaultPolicy7C543758": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "orderTable2AB11F49",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "orderTable2AB11F49",
                          "Arn",
                        ],
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "productTable02AC15CB",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PopulateTableFunctionServiceRoleDefaultPolicy7C543758",
        "Roles": [
          {
            "Ref": "PopulateTableFunctionServiceRoleB44223E7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "orderTable2AB11F49": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "lineId",
            "AttributeType": "S",
          },
          {
            "AttributeName": "email",
            "AttributeType": "S",
          },
          {
            "AttributeName": "date",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": [
          {
            "IndexName": "emailDateIndex",
            "KeySchema": [
              {
                "AttributeName": "email",
                "KeyType": "HASH",
              },
              {
                "AttributeName": "date",
                "KeyType": "RANGE",
              },
            ],
            "Projection": {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "lineId",
            "KeyType": "HASH",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "productTable02AC15CB": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
